File: katas/6 kyu/valid-braces/solution.py
# valid-braces
import codewars_test as test
# -------------------------------- MY SOLUTION ------------------------------- #
def valid_braces(string):
  brackets = ['()', '{}', '[]']
  while any(x in string for x in brackets):
    for br in brackets:
      string = string.replace(br, '')
  return not string
# ------------------------------ CLEVER SOLUTION ----------------------------- #
def validBraces(s):
  while '{}' in s or '()' in s or '[]' in s:
    s=s.replace('{}','')
    s=s.replace('[]','')
    s=s.replace('()','')
  return s==''
# ----------------------------------- TEST ----------------------------------- #
def assert_valid(string):
    test.assert_equals(valid_braces(string), True, 'Expected "{0}" to be valid'.format(string))

def assert_invalid(string):
    test.assert_equals(valid_braces(string), False, 'Expected "{0}" to be invalid'.format(string))


@test.describe("Valid Braces")
def tests():
	@test.it("sample Tests")
	def sample_tests():
		assert_valid( "()" )
		assert_invalid( "(}" )
		assert_valid( "[]" )
		assert_invalid("[(])")
		assert_valid( "{}" )
		assert_valid( "{}()[]" )
		assert_valid( "([{}])" )
		assert_invalid( "([}{])" )
		assert_valid( "{}({})[]" )
		assert_valid( "(({{[[]]}}))" )
		assert_invalid( "(((({{" )
		assert_invalid( ")(}{][" )
		assert_invalid( "())({}}{()][][" )

File: katas/6 kyu/valid-braces/solution.php
<!-- valid-braces -->
<?php
use PHPUnit\Framework\TestCase;

use function PHPUnit\Framework\any;

/* ------------------------------- MY SOLUTION ------------------------------ */
function validBraces($braces){
  while (strpos($braces, '{}') !== false || strpos($braces, '()') !== false || strpos($braces, '[]') !== false) {
    $braces = str_replace("{}" , "" , $braces);
    $braces = str_replace("()" , "" , $braces);
    $braces = str_replace("[]" , "" , $braces);
  }
  return $braces === "";
}
/* ----------------------------- CLEVER SOLUTION ---------------------------- */
function validBraces($braces){
  do {
    $braces = str_replace(['()', '[]', '{}'], '', $braces, $count);
  } while ($count);

  return empty($braces);  
}
/* ---------------------------------- TEST ---------------------------------- */
class ValidBracesTestCases extends TestCase
{
    public function testSamples() {
      $this->assertSame(true, validBraces("()"));
      $this->assertSame(false, validBraces("[(])"));
      $this->assertSame(false, validBraces("(})"));
      $this->assertSame(false, validBraces(")(}{]["));
      $this->assertSame(false, validBraces("())({}}{()][]["));
      $this->assertSame(true, validBraces("({})[({})]"));
      $this->assertSame(true, validBraces("(({{[[]]}}))"));
      $this->assertSame(true, validBraces("{}({})[]"));
    }
}